packages/core/src/application.ts

```diff
@@ -272,6 +290,53 @@ export class Application extends Context {
     return this.bind('api-spec').to(spec);
   }
 
+  /**
+   * Get the OpenAPI specification describing the REST API provided by
+   * this application.
+   *
+   * This method merges operations (HTTP endpoints) from the following sources:
+   *  - `app.api(spec)`
+   *  - `app.controller(MyController)`
+   *  - `app.route(route)`
+   *  - `app.route('get', '/greet', operationSpec, MyController, 'greet')`
+   */
+  getApiSpec(): OpenApiSpec {
+    const spec = this.getSync('api-spec');
+    this._setupHandlerIfNeeded();
+
+    // Apply deep-assign (deep clone) to prevent getApiSpec() callers from
+    // accidentally modifying our internal routing data
+    spec.paths = deepAssign({}, this._httpHandler.describeApiPaths());
```

Comment:
Based on my previous comment, I went to look at the deep-assign library, and it looks like it's only meant to perform a recursive `Object.assign` which would still run afoul of MDN's deep clone warning.
https://www.npmjs.com/package/deep-assign